### PR TITLE
Adjust how native static libraries are linked

### PR DESCRIPTION
**Pull request by retep998**
Switch from `link_whole_staticlib` to `link_staticlib` because we now pass explicit lists of what to export to the linker on all platforms so theoretically it should work.
Ensure that `hint_static` is used appropriately for all static libraries, including `static-nobundle`, not just `static`.

r? @alexcrichton 
cc @vadimcn

### CODE REVIEW

File: src/librustc_trans/back/link.rs [hunk 1]
```diff
@@ -1021,38 +1021,23 @@ fn add_local_native_libraries(cmd: &mut Linker, sess: &Session) {
...
-    for l in staticlibs {
-        // Here we explicitly ask that the entire archive is included into the
-        // result artifact. For more details see #15460, but the gist is that
-        // the linker will strip away any unused objects in the archive if we
-        // don't otherwise explicitly reference them. This can occur for
-        // libraries which are just providing bindings, libraries with generic
-        // functions, etc.
-        cmd.link_whole_staticlib(&l.name.as_str(), &search_path);
-    }
+    }).collect::<Vec<_>>();
```
Comment (frewsxcv) [inline src/librustc_trans/back/link.rs]: Any reason for the `collect`?

Comment (retep998) [inline src/librustc_trans/back/link.rs]: Not really, so I removed it.

**Comment by retep998** [comment]
Hmmmm, it looks like linking the whole staticlib is still necessary for librustc_llvm.

**Comment by alexcrichton** [comment]
Oh right, I don't think we past a list of symbols when creating dylibs on Unix platforms (only on Windows). For cdylibs I think we past the list on all platforms.

**Comment by vadimcn** [comment]
Any reason to not always use an explicit export list?   Are there linkers that do not support them?

**Comment by alexcrichton** [comment]
Unsure, if we can get it to work then we can get it to work. Very little of `src/librustc_trans/back/*.rs` is anything beyond "this was the first thing that landed"

**Comment by alexcrichton** [comment]
ping @retep998, any update on this?

**Comment by retep998** [comment]
https://github.com/rust-lang/rust/pull/40805 does the changes to hinting in a way which I think is better. It doesn't however attempt to switch from whole_staticlib to simply staticlib, so I'll investigate whether it is possible to use an explicit export list for dylibs everywhere.

**Comment by alexcrichton** [comment]
Ok, in the meantime I'm going to close this, but we can reopen/resubmit when tests are passing

**Pull request closed without merging**